test/plugin/base_test.rb

```diff
@@ -1280,6 +1305,50 @@ def verify_json_log_entries(n, params, payload_type = 'textPayload')
     assert i == n, "Number of entries #{i} does not match expected number #{n}"
   end
 
+  # Verify tags. Detailed implementation defers between the grpc and non-gprc
+  # paths, which is reflected in setup_logging_stubs_block and
+  # verify_log_name_block.
+  def verify_tag_acceptance(setup_logging_stubs_block, verify_log_name_block)
+    setup_gce_metadata_stubs
+    [
+      # When require_valid_tags is on, we only accept string tags with valid
+      # characters.
+      ['test', 'test', true],
+      ['germanß', 'german%C3%9F', true],
+      ['chinese中', 'chinese%E4%B8%AD', true],
+      ['specialCharacter/_-.', 'specialCharacter%2F_-.', true],
+
+      # When require_valid_tags is off, we try to convert invalid tags by
+      # stripping off invalid characters.
+      [123, '123', false],
+      ['test', 'test', false],
+      ['germanß', 'german%C3%9F', false],
+      ['chinese中', 'chinese%E4%B8%AD', false],
+      ['specialCharacter/_-.#@*&^', 'specialCharacter%2F_-.', false],
+      ["nonutf8#{[0x92].pack('C*')}", 'nonutf8', false],
+      [[1, 2, 3], '123', false],
+      [{ key: 'value' }, 'keyvalue', false],
+      ["abc#{[0x92].pack('C*')}", 'abc', false],
+      ["#{[0x92].pack('C*')}", '_', false],
+      ['abc@&^$*', 'abc', false],
+      ['@&^$*', '_', false]
+    ].each do |(tag, converted_tag, require_valid_tags)|
+      setup_driver_block = lambda do
+        @logs_sent = []
+        if require_valid_tags
+          d = create_driver(REQUIRE_VALID_TAGS_CONFIG, tag)
+        else
+          d = create_driver(APPLICATION_DEFAULT_CONFIG, tag)
+        end
+        d.emit('msg' => log_entry(0))
+        d.run
+      end
+      setup_logging_stubs_block.call(converted_tag, setup_driver_block)
+      verify_log_entries(1, COMPUTE_PARAMS, 'structPayload')
```

Comment:
Remember, `verify_log_entries` takes a block for extra verification, so this could instead be:
```
verify_log_entries(1, COMPUTE_PARAMS, 'structPayload') do |entry|
  assert_equal "projects/#{PROJECT_ID}/logs/#{converted_tag}",
               entry['logName']
end